File: apps/ios/PhrenKit/Tests/PhrenKitTests/SyntaxTokenizerTests.swift
import XCTest
@testable import PhrenKit

final class SyntaxTokenizerTests: XCTestCase {
    private func kinds(_ line: String, _ language: SyntaxTokenizer.Language) -> [(String, SyntaxTokenizer.Kind)] {
        SyntaxTokenizer.tokenize(line, language: language).map { (String(line[$0.range]), $0.kind) }
    }

    func testDetectsLanguagesFromPathsAndFenceLabels() {
        XCTAssertEqual(SyntaxTokenizer.Language.detect("Sources/App/Theme.swift"), .swift)
        XCTAssertEqual(SyntaxTokenizer.Language.detect("scripts/deploy-phone.py"), .python)
        XCTAssertEqual(SyntaxTokenizer.Language.detect("ts"), .typescript)
        XCTAssertEqual(SyntaxTokenizer.Language.detect("bash"), .shell)
        XCTAssertEqual(SyntaxTokenizer.Language.detect("Makefile"), .plain)
        XCTAssertEqual(SyntaxTokenizer.Language.detect(nil), .plain)
    }

    func testSwiftLineTintsKeywordsTypesCallsStringsAndComments() {
        let tokens = kinds(#"let accent = Color(hex: 0xB994F4) // the theme's purple"#, .swift)
        XCTAssertEqual(tokens.first?.0, "let"); XCTAssertEqual(tokens.first?.1, .keyword)
        XCTAssertTrue(tokens.contains { $0 == ("Color", .function) || $0 == ("Color", .type) })
        XCTAssertTrue(tokens.contains { $0 == ("0xB994F4", .number) })
        XCTAssertTrue(tokens.contains { $0.0.hasPrefix("//") && $0.1 == .comment })
        // Nothing inside the comment is re-tinted as a keyword.
        XCTAssertFalse(tokens.contains { $0 == ("the", .keyword) })
    }

    func testStringsShieldTheirContents() {
        let tokens = kinds(#"print("if this were code", 12)"#, .python)
        XCTAssertEqual(tokens.map(\.0), ["print", "\"if this were code\"", "12"])
        XCTAssertEqual(tokens.map(\.1), [.keyword, .string, .number])
    }

    func testJsonKeysAndShellCommentsAndPlain() {
        XCTAssertEqual(kinds(#"  "content": "import Foundation", "count": 3"#, .json).map(\.1), [.attribute, .string, .attribute, .number])
        XCTAssertEqual(kinds("export PATH=/usr/bin # tools", .shell).map(\.0), ["export", "# tools"])
        XCTAssertTrue(kinds("anything at all", .plain).isEmpty)
        XCTAssertTrue(SyntaxTokenizer.tokenize("", language: .swift).isEmpty)
    }

    func testTokensNeverOverlapAndAreOrdered() {
        let line = #"func run(_ name: String) throws -> [Token] { return "x\"y" }"#
        let tokens = SyntaxTokenizer.tokenize(line, language: .swift)
        for (a, b) in zip(tokens, tokens.dropFirst()) { XCTAssertLessThanOrEqual(a.range.upperBound, b.range.lowerBound) }
    }
}

File: apps/android/phrenkit/src/test/kotlin/com/phren/kit/SyntaxTokenizerTests.kt
package com.phren.kit

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import com.phren.kit.SyntaxTokenizer.Kind
import com.phren.kit.SyntaxTokenizer.Language

/** SyntaxTokenizerTests.swift, case for case. */
class SyntaxTokenizerTests {
    private fun kinds(line: String, language: Language) = SyntaxTokenizer.tokenize(line, language).map { line.substring(it.range) to it.kind }

    @Test fun detectsLanguagesFromPathsAndFenceLabels() {
        assertEquals(Language.SWIFT, Language.detect("Sources/App/Theme.swift"))
        assertEquals(Language.PYTHON, Language.detect("scripts/deploy-phone.py"))
        assertEquals(Language.TYPESCRIPT, Language.detect("ts"))
        assertEquals(Language.SHELL, Language.detect("bash"))
        assertEquals(Language.PLAIN, Language.detect("Makefile"))
        assertEquals(Language.PLAIN, Language.detect(null))
    }

    @Test fun swiftLineTintsKeywordsTypesCallsStringsAndComments() {
        val tokens = kinds("let accent = Color(hex: 0xB994F4) // the theme's purple", Language.SWIFT)
        assertEquals("let" to Kind.KEYWORD, tokens.first())
        assertTrue(tokens.any { it == ("Color" to Kind.FUNCTION) || it == ("Color" to Kind.TYPE) })
        assertTrue(tokens.any { it == ("0xB994F4" to Kind.NUMBER) })
        assertTrue(tokens.any { it.first.startsWith("//") && it.second == Kind.COMMENT })
        assertFalse(tokens.any { it == ("the" to Kind.KEYWORD) })
    }

    @Test fun stringsShieldTheirContents() {
        val tokens = kinds("print(\"if this were code\", 12)", Language.PYTHON)
        assertEquals(listOf("print", "\"if this were code\"", "12"), tokens.map { it.first })
        assertEquals(listOf(Kind.KEYWORD, Kind.STRING, Kind.NUMBER), tokens.map { it.second })
    }

    @Test fun jsonKeysAndShellCommentsAndPlain() {
        assertEquals(listOf(Kind.ATTRIBUTE, Kind.STRING, Kind.ATTRIBUTE, Kind.NUMBER), kinds("  \"content\": \"import Foundation\", \"count\": 3", Language.JSON).map { it.second })
        assertEquals(listOf("export", "# tools"), kinds("export PATH=/usr/bin # tools", Language.SHELL).map { it.first })
        assertTrue(kinds("anything at all", Language.PLAIN).isEmpty())
        assertTrue(SyntaxTokenizer.tokenize("", Language.SWIFT).isEmpty())
    }

    @Test fun tokensNeverOverlapAndAreOrdered() {
        val line = "func run(_ name: String) throws -> [Token] { return \"x\\\"y\" }"
        val tokens = SyntaxTokenizer.tokenize(line, Language.SWIFT)
        tokens.zipWithNext().forEach { (a, b) -> assertTrue(a.range.last < b.range.first) }
    }
}
